### ISSUE TITLE
how to replace fbx Models

### ISSUE THREAD

**Issue by birdflies** [nagyrajmund/gesticulating_agent_unity#7]
Dear,
Sorry to bother you again. It seems that the FBX models in the project is come from adobe Mixamo (Y bot).
But when i replace the fbx model with other models from Mixamo such as "Monsey", the bone movements become so weird?
Do you encounter the same problem, and how to solve it?
Thanks a million～

**Comment by nagyrajmund** [comment]
Hi @birdflies ,

Yes, I experienced the same issues. The problem is that different FBX models may interpret rotations differently because there is no convention on how their coordinate systems are defined.

The neural network that we use for generating the gestures produces the motion according to the Biovision BVH format (i.e. local joint rotations while assuming a right handed coordinate system with the Y axis as the world up vector). The generated rotations then had to be transformed to the coordinate system of the Y bot as follows:

```
// see MotionVisualizer.cs,  StartGesticulation() function
var rot = Quaternion.Euler(-x, y, z);
rot.w *= -1;
```

If you know the coordinate system of the FBX file, then it should be possible to define the corresponding transformation here. Otherwise it's a bit tricky, and you might have to do it by trial-and-error (e.g. by flipping the signs of the x,y,z and w parameters above).

**Comment by birdflies** [comment]
@nagyrajmund  Thank you for your quick answer.
Besides Y bot, Had you attempted other models of adobe Mixamo?
If I train our our data use gesticulator or styleGesture, do you know how to process video dataset and prepare the train dataset? 
Thanking you a lot in anticipation of your reply~

**Comment by nagyrajmund** [comment]
> Besides Y bot, Had you attempted other models of adobe Mixamo?

No, I haven't tried any other models. If you manage to get one working, please do share the necessary changes so that we can add it to the repository :) 

>  If I train our our data use gesticulator or styleGesture, do you know how to process video dataset and prepare the train dataset?

If you have a video dataset, then as a first step you need to extract the joint angles from it. If you have the joint angles in the BVH format, then both gesticulator and StyleGestures should be able to support it. I think you only have to update the joint names the [`JointSelector` transformation](https://github.com/Svito-zar/gesticulator/blob/a22c8b8ce43cadcf666fb72f0c03fe15dd0ebc3e/gesticulator/data_processing/bvh2features.py#L43), which is present in both codebases.

**Comment by nagyrajmund** [comment]
The rotations in the BVH files are converted to the [axis-angle representation](https://en.wikipedia.org/wiki/Axis%E2%80%93angle_representation) in both models, so an alternative solution would be to extract the motion in that format, and feed it directly into the models (after normalizing the values to the standard normal distribution). That way the entire BVH processing can be avoided.

**Comment by birdflies** [comment]
@nagyrajmund  Thanks, is there any relationshape between 3d point axis and axis-angle representation?

**Comment by birdflies** [comment]
> Hi @birdflies ,
> 
> Yes, I experienced the same issues. The problem is that different FBX models may interpret rotations differently because there is no convention on how their coordinate systems are defined.
> 
> The neural network that we use for generating the gestures produces the motion according to the Biovision BVH format (i.e. local joint rotations while assuming a right handed coordinate system with the Y axis as the world up vector). The generated rotations then had to be transformed to the coordinate system of the Y bot as follows:
> 
> ```
> // see MotionVisualizer.cs,  StartGesticulation() function
> var rot = Quaternion.Euler(-x, y, z);
> rot.w *= -1;
> ```
> 
> If you know the coordinate system of the FBX file, then it should be possible to define the corresponding transformation here. Otherwise it's a bit tricky, and you might have to do it by trial-and-error (e.g. by flipping the signs of the x,y,z and w parameters above).

 @nagyrajmund  [deep-motion-editing](https://github.com/DeepMotionEditing/deep-motion-editing.git) can make Motion Retargeting. Maybe the most important things we need to do is to retarget the bvh format of other models to the format of Y bot?

**Comment by nagyrajmund** [comment]
>is there any relationshape between 3d point axis and axis-angle representation?

I am not sure what you mean by '3d point axis'. If you are talking about the 3D coordinates of the joints (that is, their positions), then you can compute the local joint angles (which is used in the BVH format, for example) using the known bone lengths and trigonometry. 

Then, once you have the 3D joint angles, you can use e.g. the [transforms3d library](https://github.com/matthew-brett/transforms3d/blob/f185e866ecccb66c545559bc9f2e19cb5025e0ab/transforms3d/euler.py#L407) to convert them to the axis-angle representation. 

The difference between the two representations is that joint angles `[x_angle, y_angle, z_angle]` specify the rotations along each of the XYZ axes, while the axis-angle representation `[a, b, c] * Θ` represents single rotation of `Θ` degrees along a single arbitrary axis specified by the normalized vector `[a,b,c]`.

**Comment by nagyrajmund** [comment]
> deep-motion-editing can make Motion Retargeting. Maybe the most important things we need to do is to retarget the bvh format of other models to the format of Y bot?

There are two things that you have to consider here. One of them is the _difference between the coordinate systems_ of different FBX files - for example, it might happen that for some models the rotation vector `[90°, 0°, 0°]` will rotate joints along the X axis, while for other models it might rotate the joints along the Z axis. So the axes might be different and the rotations have to be changed accordingly. This is the problem that I talked about earlier.

The other problem that you might run into is that the skeleton in the training data might be different than the skeleton of the 3D avatar in Unity. For example, if you want to animate a monster with the default gesticulator model we provide, then you have to _retarget_ the generated motion to the desired skeleton of your monster. In that case, the linked repository will certainly be useful.

**Comment by birdflies** [comment]
@nagyrajmund  Ok,thanks for your help~

**Comment by Svito-zar** [comment]
@birdflies, can this issue be resolved now?

**Comment by birdflies** [comment]
yes. @Svito-zar  thanks~
